### PR TITLE
ct_image_reader

### LINKED ISSUE #78: CT image reader
**Is your feature request related to a problem? Please describe.**

This issue is to develop a CT image reader, which will use pydicom lib.

**Describe the solution you'd like**
this reader will output the CT image as numpy array in HU unit.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by bhaskar2443053**
This PR fixes #78 . 

### Description of changes
This a new CT image reader implementation using pydicom lib.

### Possible influences of this PR.
The sample data is a bit huge, even for one instance.

### Test Conducted
the test case verifies the existence of the image path from the reader in the source folder.

### CODE REVIEW

**Comment by bhaskar2443053** [comment]
Hi @Piyush13y and @hunterhector I am facing this new error with already merged processors.

E   AttributeError: module 'numpy' has no attribute 'object'

is there something we can do ?

**Comment by hunterhector** [comment]
> Hi @Piyush13y and @hunterhector I am facing this new error with already merged processors.
> 
> E AttributeError: module 'numpy' has no attribute 'object'
> 
> is there something we can do ?

Could you try to narrow down the error? Most likely there are some new environment upgrades that cause this. I cannot easily figure it out just looking at the error.

**Comment by bhaskar2443053** [comment]
The same error with xray processor test case

**Comment by hunterhector** [comment]
After a quick glance it looks like it is the spacy and numpy have some version conflicts. I would recommend recreating this issue locally and find out the right version combination.

To be honest, we haven't upgraded some library versions for a long time. We should spend a sprint on that alone.

**Comment by bhaskar2443053** [comment]
> After a quick glance it looks like it is the spacy and numpy have some version conflicts. I would recommend recreating this issue locally and find out the right version combination.
> 
> To be honest, we haven't upgraded some library versions for a long time. We should spend a sprint on that alone.

also got it with x-ray processor test too. it was added last month.

File: fortex/health/readers/CT_image_reader.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This class is designed to read Xray images from given image path.
+
+    For more information for the dataset, visit:
+      https://data.mendeley.com/datasets/jctsfj2sfn/1
+    """
```
Comment (hunterhector) [inline fortex/health/readers/CT_image_reader.py:18]: additional spaces

File: fortex/health/readers/CT_image_reader.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+import os
+from typing import Any, Iterator
+from forte.data.data_pack import DataPack
+from forte.data.base_reader import PackReader
+from forte.data.data_utils_io import dataset_path_iterator
+import pydicom
+
+
+class CTimageReader(PackReader):
+    r""":class:`CTimageReader` is designed to read CT image files from a given folder."""
```
Comment (hunterhector) [inline fortex/health/readers/CT_image_reader.py:28]: We should describe this more clearly, what types of data, what's the folder structure?

File: fortex/health/readers/CT_image_reader.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+    def __init__(self):
+        super().__init__()
+        self.pydicom = pydicom
```
Comment (hunterhector) [inline fortex/health/readers/CT_image_reader.py:32]: why do we need a `self.pydicom`? Can we simply us `pydicom` later?

File: fortex/health/readers/CT_image_reader.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+    def _collect(self, image_directory) -> Iterator[Any]:
```
Comment (hunterhector) [inline fortex/health/readers/CT_image_reader.py:34]: iterator type is too broad.

File: fortex/health/readers/CT_image_reader.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        return os.path.basename(image_file)
+
+    def _parse_pack(self, file_path: str) -> Iterator[DataPack]:
+        pack: DataPack = DataPack()
+
+        # Read in image data and store in DataPack
+        img = self.pydicom.dcmread(
+            file_path, **(self.configs.read_kwargs or {})
+        )
+        img.PhotometricInterpretation = "YBR_FULL"
```
Comment (hunterhector) [inline fortex/health/readers/CT_image_reader.py:55]: I think we should document these choices (why YBR_FULL)

File: fortex/health/readers/CT_image_reader.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        pixel_data = img.pixel_array
+        pack.add_image(image=pixel_data)
+        pack.pack_name = file_path
```
Comment (hunterhector) [inline fortex/health/readers/CT_image_reader.py:58]: `pack_name` is better to be a non-path string (not containing `/`)

File: setup.py [hunk 7]
```diff
@@ -31,6 +31,7 @@
             "transformers==4.18.0",
             "protobuf==3.19.4",
             "Pillow==8.4.0",
+            "pydicom==2.3.1",
```
Comment (hunterhector) [inline setup.py:34]: This shouldn't be an entry in `test`, it should have its own keyword in `extras_require`

File: tests/fortex/health/readers/CT_image_reader_test.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+        self.reader = CTimageReader()
+        self.pl = Pipeline[DataPack]()
+        self.pl.set_reader(self.reader)
+        self.pl.initialize()
+
+    def test_reader(self):
+        for pack in self.pl.process_dataset(self.orig_image_pth):
+            self.assertTrue(
+                pack.pack_name.split("/")[-1] in self.expected_image_path
```
Comment (hunterhector) [inline tests/fortex/health/readers/CT_image_reader_test.py:52]: `pack_name` isn't the place for storing paths. If we want to store path, we have discussed to use `Payload`

**Review by hunterhector — COMMENTED** [review]
See my comments inline.

An additional note on the binary files. These binary example files are very useful but they will make the master branch larger and larger. Could you add them in another PR using the following trick:

https://gist.github.com/mcroach/811c8308f4bd78570918844258841942

Basically, you can create an asset branch (you can use another name, say `data` branch) that holds the content, and download them when needed. This way we can keep the master branch lean.

File: setup.py (hunk 7 above)
Comment (bhaskar2443053) [reply setup.py]: done

File: .github/workflows/main.yml [hunk 9]
```diff
@@ -13,16 +13,15 @@ jobs:
...
+        python-version: [ 3.7, 3.8, 3.9 ]
         torch-version: [ 1.5.0, 1.6.0, 1.7.1, 1.8.1 ]
         tensorflow-version: [ 1.15.0, 2.2.0, 2.5.0 ]
         test-details:
           - { dep: forte-wrapper, testfile: tests/fortex/health/processors/icd_coding_processor_test.py }
           - { dep: forte-wrapper, testfile: tests/fortex/health/processors/negation_context_analysis_test.py }
-          - { dep: scispacy, testfile: tests/fortex/health/processors/scispacy_processor_test.py }
           - { testfile: tests/fortex/health/readers/mimic3_note_reader_test.py }
           - { testfile: tests/fortex/health/readers/xray_image_reader_test.py }
-          - { testfile: tests/fortex/health/processors/xray_processor_test.py }
```
[the remark this answers is not in the document]
Comment (bhaskar2443053) [reply .github/workflows/main.yml:25]: resolved with numpy version downgrade

File: fortex/health/readers/CT_image_reader.py (hunk 3 above)
Comment (bhaskar2443053) [reply fortex/health/readers/CT_image_reader.py]: I have removed the self.pydicom and used pydicom directly. 

resolved

File: .github/workflows/main.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (bhaskar2443053) [reply .github/workflows/main.yml:25]: @hunterhector I don't have access to create data branch in Forte repo. Either you can create a branch or give me the necessary access.

[the remark this answers is not in the document]
Comment (hunterhector) [reply .github/workflows/main.yml:25]: I have created a complete empty data branch: https://github.com/asyml/forte/tree/data

**Pull request closed without merging**